lib/hsa/hsaPulsarUpdatePhase.cpp

```diff
@@ -129,6 +114,41 @@ int hsaPulsarUpdatePhase::wait_on_precondition(int gpu_frame_id) {
         return -1;
     metadata_buffer_precondition_id =
         (metadata_buffer_precondition_id + 1) % metadata_buf->num_frames;
+
+
+    // Wait for new gain
+    DEBUG("Waiting for gain_buf_id={:d} to be full; gpu_frame_id={:d}", gain_buf_precondition_id,
+          gpu_frame_id);
+    if (first_pass) {
+        uint8_t* frame =
+            wait_for_full_frame(gain_buf, unique_name.c_str(), gain_buf_precondition_id);
+        gain_buf_precondition_id = (gain_buf_precondition_id + 1) % gain_buf->num_frames;
+        first_pass = false;
+        frame_to_fill = gain_buf->num_frames;
+        frame_to_fill_finalize = frame_to_fill;
+        filling_frame = true;
+        if (frame == NULL)
+            return -1;
+    } else {
+        // Check for new gains only if filled all gpu frames (not currently filling frame)
+        if (!filling_frame) {
+            auto timeout = double_to_ts(0);
+            int status = wait_for_full_frame_timeout(gain_buf, unique_name.c_str(),
+                                                     gain_buf_precondition_id, timeout);
+            DEBUG("status of gain_buf_precondition_id[{:d}]={:d} ==(0=ready 1=not)",
+                  gain_buf_precondition_id, status);
+            if (status == 0) {
```

Comment:
I don't think you are doing the right thing here.  You always copy in gains*pointings in this command object, so you don't need to copy the gains to the GPU.  All you need to do is update the gains used to the compute the pointings and then trigger an "update" to have it compute new pointings with the updated gains and switch banks.

Reply:
i have put the warning in `update_gains_frb_callback` and `update_gains_psr_callback`. is that OK?